src/common/ceph_timer.h

```diff
@@ -15,320 +15,300 @@
-      }
-
-      // Suspend operation of the timer (and let its thread die).
-      void suspend() {
-	unique_lock l(lock);
-	if (suspended)
-	  return;
-
-	suspended = true;
-	cond.notify_one();
-	l.unlock();
-	thread.join();
-      }
-
-
-      // Resume operation of the timer. (Must have been previously
-      // suspended.)
-      void resume() {
-	unique_lock l(lock);
-	  if (!suspended)
-	  return;
-
-	suspended = false;
-	ceph_assert(!thread.joinable());
-	thread = std::thread(&timer::timer_thread, this);
-      }
-
-      // Schedule an event in the relative future
-      template<typename Callable, typename... Args>
-      uint64_t add_event(typename TC::duration duration,
-			 Callable&& f, Args&&... args) {
-	typename TC::time_point when = TC::now();
-	when += duration;
-	return add_event(when,
-			 std::forward<Callable>(f),
-			 std::forward<Args>(args)...);
-      }
+#include "include/function2.hpp"
 
-      // Schedule an event in the absolute future
-      template<typename Callable, typename... Args>
-      uint64_t add_event(typename TC::time_point when,
-			 Callable&& f, Args&&... args) {
-	std::lock_guard l(lock);
-	event& e = *(new event(
-		       when, ++next_id,
-		       std::forward<std::function<void()> >(
-			 std::bind(std::forward<Callable>(f),
-				   std::forward<Args>(args)...))));
-	auto i = schedule.insert(e);
-	events.insert(e);
-
-	/* If the event we have just inserted comes before everything
-	 * else, we need to adjust our timeout. */
-	if (i.first == schedule.begin())
-	  cond.notify_one();
-
-	// Previously each event was a context, identified by a
-	// pointer, and each context to be called only once. Since you
-	// can queue the same function pointer, member function,
-	// lambda, or functor up multiple times, identifying things by
-	// function for the purposes of cancellation is no longer
-	// suitable. Thus:
-	return e.id;
-      }
-
-      // Adjust the timeout of a currently-scheduled event (relative)
-      bool adjust_event(uint64_t id, typename TC::duration duration) {
-	return adjust_event(id, TC::now() + duration);
-      }
-
-      // Adjust the timeout of a currently-scheduled event (absolute)
-      bool adjust_event(uint64_t id, typename TC::time_point when) {
-	std::lock_guard l(lock);
-
-	event key(id);
-	typename event_set_type::iterator it = events.find(key);
-
-	if (it == events.end())
-	  return false;
-
-	event& e = *it;
-
-	schedule.erase(e);
-	e.t = when;
-	schedule.insert(e);
-
-	return true;
-      }
-
-      // Cancel an event. If the event has already come and gone (or you
-      // never submitted it) you will receive false. Otherwise you will
-      // receive true and it is guaranteed the event will not execute.
-      bool cancel_event(const uint64_t id) {
-	std::lock_guard l(lock);
-	event dummy(id);
-	auto p = events.find(dummy);
-	if (p == events.end()) {
-	  return false;
-	}
-
-	event& e = *p;
-	events.erase(e);
+namespace ceph {
+namespace bi = boost::intrusive;
```

Comment:
this causes a name colision:
```
src/osd/Watch.cc
In file included from /build/ceph-15.1.0-1850-g2979e56/src/osd/OSD.h:57:0,
                 from /build/ceph-15.1.0-1850-g2979e56/src/osd/Watch.cc:9:
/build/ceph-15.1.0-1850-g2979e56/src/common/WeightedPriorityQueue.h:54:29: error: reference to 'bi' is ambiguous
     class ListPair : public bi::list_base_hook<>
                             ^~
/build/ceph-15.1.0-1850-g2979e56/src/common/WeightedPriorityQueue.h:26:32: note: candidates are: namespace bi = boost::boost::intrusive;
 namespace bi = boost::intrusive;
                                ^
In file included from /build/ceph-15.1.0-1850-g2979e56/src/osd/OSD.h:28:0,
                 from /build/ceph-15.1.0-1850-g2979e56/src/osd/Watch.cc:9:
/build/ceph-15.1.0-1850-g2979e56/src/common/ceph_timer.h:30:32: note:                 namespace ceph::bi = boost::boost::intrusive;
 namespace bi = boost::intrusive;
                                ^
```

see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=bionic,DIST=bionic,MACHINE_SIZE=huge/37875//consoleFull

Reply:
Well, fudge. Razzin' frazzin' using namespace. From hell's heart I spit my last breath at it.

Fixed, pushed.